### DISCUSSION TITLE
Whether to consider exporting the data parsed by ezdxf into json for rendering using threejs on the web side

### DISCUSSION THREAD

**Discussion by LuckDogHu** [mozman/ezdxf#1039] [Ideas]
Although there is already a library for rendering dxf on npm, it is not as comprehensive as the data parsed by ezdxf, so I want to use ezdxf to parse the data on the back end, return the data to the front end through network requests and then use threejs to render. As a further idea, ezdxf internally supports importing and exporting json to support editing and modifying dxf on the web side. I am very grateful to the author of exdxf, he helped me a lot in parsing dxf.

**Comment by mbway** [comment]
This is certainly something that is possible with the drawing addon. By implementing a custom rendering backend it would be possible to render to a custom data format instead of an image. This would be relatively simple to achieve for your own needs (see the [SVG backend](https://github.com/mozman/ezdxf/blob/a9ec4151e7293cbc4dccdf45614b8e7e0e4d48c7/src/ezdxf/addons/drawing/svg.py#L168) for example).

For general use (if you would like to contribute the rendering backend upstream), a format like [geojson](https://geojson.org) would be appropriate for the use case you are describing I think.

**Comment by mozman** [comment]
There is no standard for converting DXF files to JSON. I implemented the simplest possible conversion scheme:

DXF Content:
```Text
  0
SECTION
  2
HEADER
  9
$ACADVER
  1
AC1027
  9
$ACADMAINTVER
 70
105
  9
$DWGCODEPAGE
  3
ANSI_1252
  9
$LASTSAVEDBY
  1
ezdxf
  9
$REQUIREDVERSIONS
160
0
  9
$INSBASE
 10
0.0
 20
0.0
 30
0.0
...
  0
EOF
```

JSON:
```json
[
[0, "SECTION"],
[2, "HEADER"],
[9, "$ACADVER"],
[1, "AC1032"],
[9, "$ACADMAINTVER"],
[90, "105"],
[9, "$DWGCODEPAGE"],
[3, "ANSI_1252"],
[9, "$LASTSAVEDBY"],
[1, "ezdxf"],
[9, "$REQUIREDVERSIONS"],
[160, "0"],
[9, "$INSBASE"],
[10, "0.0"],
[20, "0.0"],
[30, "0.0"],
["... many tags ..."],
[0, "EOF"]
]
```

This is a true 1:1 DXF to JSON conversion, but I doubt any application can interpret this. Once you start exporting DXF elements as objects  with named attributes, you create a new data format that no one other than your application can interpret.

The JSON exporter already exists. If this JSON format is helpful I can also implement an loader for this format.

@mbway The `geojson` backend is a good idea!

**Reply by LuckDogHu** [reply]
Thank you very much for your reply, maybe it was my fault. What I actually want is something like a definition file. For easy use in the web, geojson is a good way, but it is not good enough. I personally think that entities in dxf and geojson are still different. For example, there is the arc entity in dxf. But there is no concept of curve in gis. Json can be regarded as a bridge between python and nodejs. It's just that I want to convert objects in python into objects in nodejs through json.

in python
```python
res = []
doc = ezdxf.readfile('./文本.dxf')
msp = doc.modelspace()
for text in msp.query("TEXT"):
    text_obj = {}
    text_obj['handle'] = text.dxf.handle
    text_obj['insert'] = [text.dxf.insert.x, text.dxf.insert.y]
    res.append(text_obj)
print(json.dumps(obj=res, ensure_ascii=False))
# [{"handle": "2C5", "insert": [533575.3122716594, 3121963.351565434]}]
```
in javascript
```javascript
const arr = JSON.parse('[{"handle": "2C5", "insert": [533575.3122716594, 3121963.351565434]}]');
console.log(arr) // [ { handle: '2C5', insert: [ 533575.3122716594, 3121963.351565434 ] } ]
```
Just the above conversion is required. Not from source file to json. My original intention was to take advantage of the powerful analytical capabilities of ezdxf. To obtain the information that has been extracted. Do something with the parsed data.

**Reply by mozman** [reply]
OK, that would require an extra export method for each entity because not all DXF attributes are simple values e.g. LWPOLYLINE vertices.

**Reply by LuckDogHu** [reply]
Yes. My thought is that the two can be inverse of each other. That is to say. I can export a json definition file, and I can restore dxf based on this json. This is just an idea I had. Maybe you are a little confused about English (because my English is very poor, I can only translate it through google. I'm sorry for the inconvenience) Thank you！ 🙏

**Reply by mozman** [reply]
Ok I understand. This can be an add-on and does not have to be included in the core library. I have no interest and no time to implement this, but I would accept such a contribution as an add-on 😊.

**Comment by mozman** [comment]
I added two new backends to the `drawing` add-on.

## The `GeoJSONBackend` outputs JSON data according the [GeoJSON](https://geojson.org/) specs.

GeoJSON uses a geographic coordinate reference system, World Geodetic System 1984, and units of decimal degrees.

- Latitude: -90 to +90 (South/North)

- Longitude: -180 to +180 (East/West)

So most DXF files will produce invalid coordinates and it is the job of the package-user to transform the content accordingly!
The GeoJSON format supports only straight lines so curved shapes are flattened to polylines and polygons.

Data loaded and displayed by https://geojson.io (the location of the source DXF file is wrong):
![image](https://github.com/mozman/ezdxf/assets/4230966/cc21c20b-5cd7-4884-9156-ea58ef76d444)

I don't think this backend very useful because of the coordinate constrains but is was fun and wasn't that much work.

## A `CustomJSONBackend` with a custom JSON scheme. 

This scheme supports curved shapes by a SVG-path like structure and coordinates are not limited in any way. This backend can be used to send geometries from a web-backend to a frontend.

The JSON scheme is documented in the source code:

https://github.com/mozman/ezdxf/blob/master/src/ezdxf/addons/drawing/json.py

This backend is only tested in theory by converting the CADKit samples to JSON, but there is no viewer to display the result.
